Product subscriptions: Free trial section

0nko opened the pull request:
Implements #10148, which adds a new Subscription free trial to the product detail and variation detail screen.

| Product detail | Subscription free trial |
| --- | --- |
| ![image](https://github.com/woocommerce/woocommerce-android/assets/1522856/38fcf858-4981-4390-a984-51a7090ea20e) | ![image](https://github.com/woocommerce/woocommerce-android/assets/1522856/055b5ea1-1790-49c2-9ec3-da9b24b0d91c) |

**To test:**

Simple subscription product
1. Make sure you have the Subscriptions plugin installed
2. Prepare a simple subscription product
3. Start the app
4. Go to Products -> Tap on the simple subscription product
5. Notice there's a new Subscription free trial section with the correct data shown
6. Tap on the free trial property
7. Notice a detail free trial screen is shown with the correct data
8. Try changing the values
9. Tap back
10. Notice the free trial section information was update

Variable subscription product
1. Make sure you have the Subscriptions plugin installed
2. Prepare a variable subscription product (make sure you generate some variations)
3. Start the app
4. Go to Products -> Tap on the variable subscription product
5. Tap on Variations -> Select a variation
6. Notice there's a new Subscription free trial section with the correct data shown
7. Tap on the free trial property
8. Notice a detail free trial screen is shown with the correct data
9. Try changing the values
10. Tap back
11. Notice the free trial section information was update

0nko committed:
Add free trial resources

0nko committed:
Add a new navigation target

0nko committed:
Add a trial period display extension function

0nko committed:
Add new subscription trial product property

0nko committed:
Add variation subscription property

0nko committed:
Add a background to the expiration screen

0nko committed:
Add free trial fragment and navigation

0nko committed:
Replace exit classes with no properties with objects

0nko committed:
Move the free trial view logic into a ViewModel

0nko committed:
Handle product detail navigation and result handling

0nko committed:
Handle variation detail navigation and result handling

0nko committed:
Optimize imports

0nko committed:
Add a free trial screen preview

0nko committed:
Remove unused class

0nko committed:
Add unit tests

0nko commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/subscriptions/ProductSubscriptionExpirationFragment.kt:
```diff
@@ -80,28 +87,45 @@ class ProductSubscriptionExpirationFragment : BaseProductFragment() {
         items: List<String>,
         currentValue: String
     ) {
-        Box(
```
This is unrelated to the free trial but adds the grey background to the subscription screen for consistency.

hichamboushaba replied:
Thanks for this @0nko, but I have a minor remark the usage of the `Divider` below, in our current screens, it's not a divider, but a shadow of the Card that holds the content, we can follow the same if we want, WDYT? It's a minor point.

Just for information, I experienced with this layout:
```kotlin
Box(Modifier.fillMaxSize()) {
            Card(
                shape = RectangleShape,
                modifier = Modifier
                    .fillMaxWidth()
            ) {
                Row(
                    modifier = Modifier
                        .padding(dimensionResource(id = dimen.major_100)),
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    OutlinedTextField(
                        value = if (length == 0) "" else length.toString(),
                        onValueChange = { input ->
                            val trialLength = if (input.isNotEmpty()) {
                                input.filter { it.isDigit() }.toInt()
                            } else {
                                0
                            }
                            onTrialLengthUpdated(trialLength)
                        },
                        keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number),
                        modifier = Modifier
                            .weight(.3f)
                            .padding(end = dimensionResource(id = dimen.major_100))
                    )
                    WcExposedDropDown(
                        items = items.toTypedArray(),
                        onSelected = { onTrialPeriodUpdated(it) },
                        currentValue = period,
                        mapper = { it.getPeriodString(resourceProvider, length) },
                        modifier = Modifier
                            .weight(.7f)
                            .background(colorResource(id = color.color_surface))
                            .padding(start = dimensionResource(id = dimen.major_100))
                    )
                }
            }
        }
```

and it gives the following:
<img width=360 src="https://github.com/woocommerce/woocommerce-android/assets/1657201/d63ae075-b9b7-4153-9548-02d7adecccc6"/>

0nko replied:
Thanks for the suggestion! You're right about this. I copied what I saw it some layout XML but this is the right approach 👍.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailViewModel.kt:
```diff
@@ -767,49 +789,49 @@ class ProductDetailViewModel @Inject constructor(
         var eventName: AnalyticsEvent? = null
         var hasChanges = false
         when (event) {
-            is ProductExitEvent.ExitSettings -> {
```
np, I think these changes were an accidental refactoring by the IDE, scoping the events to the parent class `ProductExitEvent` is easier to read IMO, so I suggest reverting it.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductNavigator.kt:
```diff
@@ -314,8 +322,8 @@ class ProductNavigator @Inject constructor() {
-                        NavOptions.Builder()
-                            .setPopUpTo(R.id.dashboard, false)
+                        Builder()
+                            .setPopUpTo(id.dashboard, false)
```
Same as above, I think this was an accidental refactoring by the IDE, `NavOptions.Builder` is clearer than just `Builder`.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductNavigator.kt:
```diff
@@ -409,36 +417,44 @@ class ProductNavigator @Inject constructor() {
                 )
             }
 
-            is ProductNavigationTarget.ViewProductSubscription -> {
```
Same as above 🙂

hichamboushaba reviewed (commented):
Nice work @0nko, this works well.

I left some comments, mostly minor about IDE auto refactoring that messed up some imports.
I'm not approving yet as I see we need to implement some client side validation p1700049601331839-slack-C03L1NF1EA3, if you are planning on adding on a separate PR, please let me know.

0nko committed:
Add trial period range validation

0nko committed:
Optimize imports

0nko committed:
Add qualifier names back for better readability

0nko committed:
Add a missing import

0nko committed:
Use a shadow instead of a divider

0nko commented:
Thank you for the review @hichamboushaba! I've addressed the feedback and added range validation. Ready for another round.

<image src="https://github.com/woocommerce/woocommerce-android/assets/1522856/c7ef2825-5274-424e-b554-cdfdc867d8d1" width="300" />

0nko committed:
Merge branch 'trunk' into issue/10148-free-trial

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/VariationDetailViewModel.kt

hichamboushaba committed:
Merge branch 'trunk' into issue/10148-free-trial

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailCardBuilder.kt
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/subscriptions/ProductSubscriptionExpirationFragment.kt
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/subscriptions/SubscriptionProductUiExtensions.kt

hichamboushaba committed:
Revert unwanted import changes

hichamboushaba reviewed (approved):
Nice job @0nko, :shipit: 

Since you are AFK today, I'll fix conflicts then merge the PR

hichamboushaba committed:
Fix build issue

0nko commented:
Thank you @hichamboushaba!
